test/test_linalg.py

```diff
@@ -751,6 +751,155 @@ def test_det(self, device, dtype):
+    def test_det_autograd(self, device, dtype):
+        # Create a batch of randomized square singular matrices. Each matrix is
+        # generated by performing a matrix multiply of two random matrices of
+        # size (size[-1], 1) and (1, size[-1])
+        def make_singular_matrix_batch(size):
+            assert size[-1] == size[-2]
+
+            batch = torch.empty(size, device=device, dtype=dtype)
+
+            matrix_size = size[-2:]
+            view_size = ((1,) + size) if len(size) == 1 else ((int(np.prod(size[:-2])),) + matrix_size)
+            batch_view = batch.view(view_size)
+
+            a_size = (size[-1], 1)
+            b_size = (1, size[-1])
+
+            for matrix_idx in range(batch_view.size(0)):
+                # TODO: linalg_det_backward can sometimes produce NaNs if any
+                # singular values are 0. This is likely due to its dependence
+                # upon svd_backward. Therefore, we must avoid testing such
+                # matrices. Once lu_backward is available, linalg_det_backward
+                # can be updated to use it so that wo no longer have this
+                # problem: https://github.com/pytorch/pytorch/issues/53364
+                svd_has_zero = True
+
+                # Note: When calculating the determinant, floating point errors
+                # can be introduced such that the calculated determinant is
+                # a very small number, rather than exactly zero. We use random
+                # integers cast to float to drastically decrease the chance of
+                # this happening and then double check that the determinant is
+                # exactly zero. If determinant is nonzero, we regenerate
+                # a limited number of times until it is zero.
+                det_is_nonzero = True
+
+                num_tries = 0
+                max_tries = 50
+                matrix = None
+
+                while det_is_nonzero or svd_has_zero:
+                    num_tries += 1
+                    self.assertTrue(
+                        num_tries < max_tries,
+                        f'could not create singular matrix of size {matrix_size}'
+                        f' after trying {max_tries} times')
+
+                    lo = -5
+                    hi = 5
+                    if dtype.is_complex:
+                        a = torch.randint(lo, hi, a_size, device=device).to(dtype).add(
+                            1j * torch.randint(lo, hi, a_size, device=device).to(dtype))
+                        b = torch.randint(lo, hi, b_size, device=device).to(dtype).add(
+                            1j * torch.randint(lo, hi, b_size, device=device).to(dtype))
+
+                        # Avoiding zeros in a and b decreases likelihood of
+                        # generating a matrix with a singular value equal to 0
+                        a[a.eq(0)] = random.choice([hi, 1j * hi])
+                        b[b.eq(0)] = random.choice([hi, 1j * hi])
```

Comment:
I am not sure I agree with the statement.

Also, we have a control over which matrices we want to generate from svd directly, that is:
```
x = ... a random matrix.
u, _, v = x.svd() # we need u, v which are orthogonal/hermitian. QR, for example, could also be used instead of SVD.

# now we can just generate a tensor of singular values with any desired properties

#  a well-conditioned matrix with det(matrix) != 0
s = torch.randn(shape of s) * 0.0001 + 1.0 (singular values sampled from a gaussian with the mean 1.0 and the variance 0.0001)
matrix = (u * s.unsqueeze(-2)) @ v.transpose(-1, -2)

# a matrix of rank k. if k < n, det(matrix) is zero
s = torch.rand(shape of s, ...)
s[..., -k:] = 0
matrix = (u * s.unsqueeze(-2)) @ v.transpose(-1, -2)
```

Reply:
It would be cool to generate the matrices from svd if it makes things simpler and more easily controllable. However, there are some challenges that I'm not sure about.

I suppose there's an additional step required for generating a matrix with det == 0--we need to make at least two rows or columns are linearly dependent to downgrade the rank. Otherwise, the `s[..., -k:] = 0` just sets all the singular values to 0 and the output is just a zero matrix.

There also may still be a problem with setting singular values to 0. As mentioned in my comments, I saw a lot of NaNs when I put matrices with singular values equal to 0 into `gradcheck`. I'm not exactly sure if that was actually the cause of the NaNs though, it could potentially be a coincidence.

Finally, this method still has issues with floating point error leading to a nonzero determinant. Even if all the intermediate numbers I generate are integers, we still don't get det == 0:

```python
>>> x = torch.randint(-5, 5, (4, 4)).to(torch.cdouble) + 1j * torch.randint(-5, 5, (4, 4)).to(torch.cdouble)
>>> x[0] = x[1]
>>> k = torch.matrix_rank(x)
>>> k
tensor(3)
>>> u, _, v = x.svd()
>>> s = torch.randint(1, 10, (4,)).to(torch.cdouble) + 1j * torch.randint(1, 10, (4,)).to(torch.cdouble)
>>> s[-k:] = 0
>>> matrix = (u * s.unsqueeze(-2)) @ v.transpose(-1, -2).conj()
>>> matrix.det()
tensor(-4.0203e-47+1.1162e-46j, dtype=torch.complex128)
```

Reply:
I decided to make a script to test out different methods of creating matrices with det=0: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/det_autograd/gen_det0_mats.py

By default, it will try out each method 10,000 times for a 4x4 complex128 matrix. Then it reports what ratio of matrices had a determinant of exactly 0, and the ratio that contained a singular value of 0. Below is what I get when I run it.

```
| func_name | ratio det is exactly 0 | ratio singular value 0 |
| --- | --- | --- |
| gen_with_randn | 0.0 | 0.0 |
| gen_with_svd | 0.0792 | 0.001 |
| gen_with_matmul_floats | 0.0213 | 0.0006 |
| gen_with_matmul_ints | 0.8029 | 0.0626 |
| gen_with_matmul_ints_avoid_svd_eq_0 | 0.7906 | 0.0281 |
```

`gen_with_randn` is plain old `torch.randn`. As expected, this will almost never produce a det=0 matrix.

`gen_with_svd` is basically what I posted in my last comment (based on your previous post). It only generates a det=0 matrix about 8% of the time. Perhaps that could be improved somehow.

`gen_with_matmul_floats` is the vector-vector matmul method on tensors generated with `torch.randn`. Unsurprisingly, this doesn't give many det=0 matrices.

`gen_with_matmul_ints` is similar to the previous one, except that `torch.randint`is used to drastically increase the likelihood of det=0, up to 80%. About 6% of the time, we get a matrix containing a singular value of 0, which is a low enough probability that we could just regenerate when we encounter it.

`gen_with_matmul_ints_avoid_svd_eq_0` is the same as the previous one, except that it replaces any values equal to 0 in the two intermediate vectors (which will be matmul'd) with a nonzero value. It performs pretty much the same as the previous method, but with a little bit of an improvement regarding singular values of 0.

Now that we have a good testbench for these matrix generation methods, we can try out more ideas and quickly find out if they're feasible. Please let me know if you have any other suggestions, @nikitaved

Reply:
> I suppose there's an additional step required for generating a matrix with det == 0--we need to make at least two rows or columns are linearly dependent to downgrade the rank. Otherwise, the `s[..., -k:] = 0` just sets all the singular values to 0 and the output is just a zero matrix.

The number of non-zero singular value is one of the definitions of rank.

> 
> There also may still be a problem with setting singular values to 0. As mentioned in my comments, I saw a lot of NaNs when I put matrices with singular values equal to 0 into `gradcheck`. I'm not exactly sure if that was actually the cause of the NaNs though, it could potentially be a coincidence.

I see. that could be the issue with `svd_backward` as the underlying assumption is to have distinct singular values.
> 
> Finally, this method still has issues with floating point error leading to a nonzero determinant. Even if all the intermediate numbers I generate are integers, we still don't get det == 0:
> 
> ```python
> >>> x = torch.randint(-5, 5, (4, 4)).to(torch.cdouble) + 1j * torch.randint(-5, 5, (4, 4)).to(torch.cdouble)
> >>> x[0] = x[1]
> >>> k = torch.matrix_rank(x)
> >>> k
> tensor(3)
> >>> u, _, v = x.svd()
> >>> s = torch.randint(1, 10, (4,)).to(torch.cdouble) + 1j * torch.randint(1, 10, (4,)).to(torch.cdouble)
> >>> s[-k:] = 0
> >>> matrix = (u * s.unsqueeze(-2)) @ v.transpose(-1, -2).conj()
> >>> matrix.det()
> tensor(-4.0203e-47+1.1162e-46j, dtype=torch.complex128)
> ```

Makes sense. Should not be an issue once we have `lu_backward` for that.

Reply:
Also, the source of instabilities is the `prod` itself. Sometimes it is better to replace `prod` with `sum` in a log-space....

Reply:
Maybe it would be best to just wait for `lu_backward`

Reply:
I think it is still worth having this PR because we can fix the wrong back are for singular cases. So, it is not just a complex support, but also a big fix :)